### PR TITLE
Fix asan bug

### PR DESCRIPTION
**Pull request by szadam**
### Description

Fixes: https://github.com/oneapi-src/unified-memory-framework/issues/581
### Checklist

- [x] Code compiles without errors locally
- [x] All tests pass locally
- [x] CI workflows execute properly

### CODE REVIEW

File: src/base_alloc/base_alloc_linux.c [hunk 1]
```diff
@@ -19,8 +19,13 @@ static UTIL_ONCE_FLAG Page_size_is_initialized = UTIL_ONCE_FLAG_INIT;
...
 void *ba_os_alloc(size_t size) {
-    return mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS,
+    void *ptr = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS,
                 -1, 0);
+    // this should be unnecessary but pairs of mmap/munmap do not reset
+    // asan's user-poisoning flags, leading to invalid error reports
+    // Bug 81619: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=81619
+    utils_annotate_memory_defined(ptr, size);
+    return ptr;
 }
```
Comment (lukaszstolarczuk) [inline src/base_alloc/base_alloc_linux.c:29]: should we also re-enable the sanitizer test?

**Review by lukaszstolarczuk — COMMENTED** [review]
pls be more descriptive in the commit msg

File: src/base_alloc/base_alloc_linux.c (hunk 1 above)
Comment (szadam) [reply src/base_alloc/base_alloc_linux.c:29]: Done

File: src/base_alloc/base_alloc_linux.c [hunk 2]
```diff
@@ -19,8 +19,13 @@ static UTIL_ONCE_FLAG Page_size_is_initialized = UTIL_ONCE_FLAG_INIT;
...
 
 void *ba_os_alloc(size_t size) {
-    return mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS,
-                -1, 0);
+    void *ptr = mmap(NULL, size, PROT_READ | PROT_WRITE,
+                     MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+    // this should be unnecessary but pairs of mmap/munmap do not reset
+    // asan's user-poisoning flags, leading to invalid error reports
+    // Bug 81619: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=81619
+    utils_annotate_memory_defined(ptr, size);
```
Comment (kswiecicki) [inline src/base_alloc/base_alloc_linux.c:27]: In this case, shouldn't this memory be marked with `utils_annotate_memory_inaccessible` in `ba_os_free` and `os_munmap`?

Comment (szadam) [reply src/base_alloc/base_alloc_linux.c:27]: In other places where we call ba_os_free, we mark the memory with utils_annotate_memory_inaccessible. The annotations are included in this function due to the known bug in ASan's user-poisoning flags, which does not reset properly with pairs of mmap/munmap. @lplewa, what do you think?

Comment (lplewa) [reply src/base_alloc/base_alloc_linux.c:27]: We can add them - it's look like there is problem with an asan marking correctly pages on mmap/unmap so we can do it on our side.

**Review by bratpiorka — CHANGES_REQUESTED** [review]
please add utils_annotate_memory_inaccessible calls according to kswiecicki comment

File: src/base_alloc/base_alloc_linux.c (hunk 2 above)
Comment (szadam) [reply src/base_alloc/base_alloc_linux.c:27]: Done

**Review by kswiecicki — APPROVED** [review]

File: src/base_alloc/base_alloc_linux.c (hunk 2 above)
Comment (kswiecicki) [reply src/base_alloc/base_alloc_linux.c:27]: Due to ASAN errors, I've asked @szadam to revert those changes.

Comment (szadam) [reply src/base_alloc/base_alloc_linux.c:27]: Okey. https://github.com/oneapi-src/unified-memory-framework/actions/runs/10039436558/job/27744736615?pr=610

**Review by lplewa — APPROVED** [review]

**Review by bratpiorka — APPROVED** [review]

**Pull request merged**